rust: Add dns support

cathay4t opened the pull request:
Due to lacking support of `auto_dns: false`, currently, only support using
static default gateway interface as DNS interface.

Rust code will try to use valid(ip enabled and manual dns allowed)
interface from desire state first, then search static gateway. I(Gris)
think this provides the small opportunity for user to enforce a interface
to hold the DNS configuration regardless the static gateway preference.

Integration test case enabled.

Changed one test case `test_add_non_canonicalized_ipv6_nameserver`:

    The test expect nmstate to
    compress `0000:0000:0000:0000:0000:0000:0000:0100` to `::100`.

    This IPv6 is IPv4-Compatible IPv6 Address(RFC 4291), so it is also
    legal to show as `::0.0.1.0` which is used by rust `std::net::Ipv6Addr`.

    Hence change the test IP to 2000::100 which is not ambiguous.

ffmancera commented on .gitignore:
```diff
@@ -28,3 +28,5 @@ rust/Cargo.lock
 rust/target
 rust/src/clib/test/nmstate_test
 rust/src/clib/test/vgcore*
+
+*.swp
```
It seems unrelated and specific to Vim. Please, put this on a different commit.

cathay4t replied:
Removed.

ffmancera commented on rust/src/cli/ncl.rs:
```diff
@@ -141,8 +141,9 @@ fn gen_conf(file_path: &str) -> Result<String, CliError> {
 #[derive(Clone, Debug, PartialEq, Serialize)]
 struct SortedNetworkState {
-    routes: Routes,
+    dns: DnsState,
     rules: RouteRules,
+    routes: Routes,
```
Does the order matter for some reason?

cathay4t replied:
So that `nmstatectl-rust` is showing the same order as `nmstatectl-python`.

ffmancera commented on rust/src/lib/dns.rs:
```diff
@@ -0,0 +1,551 @@
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct DnsClientSetting {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub server: Option<Vec<String>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub search: Option<Vec<String>>,
+    #[serde(skip)]
+    // Lower is better
+    pub(crate) priority: Option<i32>,
```
What about something like `DnsClientState`? It seems strange to me to talk about Settings on `src/lib/dns.rs`..

cathay4t replied:
Changed to `DnsClientState`

ffmancera commented on tests/integration/dns_test.py:
```diff
@@ -33,7 +33,7 @@
 EXTRA_IPV4_DNS_NAMESERVER = "9.9.9.9"
 IPV6_DNS_NAMESERVERS = ["2001:4860:4860::8888", "2606:4700:4700::1111"]
-IPV6_DNS_LONG_NAMESERVER = ["0000:0000:0000:0000:0000:0000:0000:0100"]
+IPV6_DNS_LONG_NAMESERVER = ["2000:0000:0000:0000:0000:0000:0000:0100"]
```
Why is this change required?

cathay4t replied:
Explained in commit message and also PR comment 0.

The test expect nmstate to
compress `0000:0000:0000:0000:0000:0000:0000:0100` to `::100`.

This IPv6 is IPv4-Compatible IPv6 Address(RFC 4291), so it is also
legal to show as `::0.0.1.0` which is used by rust `std::net::Ipv6Addr`.

Hence change the test IP to 2000::100 which is not ambiguous.

ffmancera replied:
ACK, thanks!

ffmancera reviewed (approved):
LGTM!

ffmancera commented:
There is a trivial lint error, could you take a look? Other than that, it looks good :-)

cathay4t committed:
rust: Add dns support

Due to lacking support of `auto_dns: false`, currently, only support
using static default gateway interface as DNS interface.

Rust code will try to use valid(ip enabled and manual dns allowed)
interface from desire state first, then search static gateway. I(Gris)
think this provides the small opportunity for user to enforce a
interface to hold the DNS configuration regardless the static gateway
preference.

Integration test case enabled.

Changed one test case `test_add_non_canonicalized_ipv6_nameserver`:

    The test expect nmstate to
    compress `0000:0000:0000:0000:0000:0000:0000:0100` to `::100`.

    This IPv6 is IPv4-Compatible IPv6 Address(RFC 4291), so it is also
    legal to show as `::0.0.1.0` which is used by rust `std::net::Ipv6Addr`.

    Hence change the test IP to 2000::100 which is not ambiguous.

Signed-off-by: Gris Ge <fge@redhat.com>
